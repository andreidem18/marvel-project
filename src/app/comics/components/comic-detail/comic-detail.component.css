.comic-detail {
    max-width: 854px;
    max-height: 700px;
    overflow-y: auto;
}

.img-loader {
    height: 300px;
    background: rgba(0, 0, 0, .2);
}

.img-loader img {
    width: 100px;
}

.btn-close {
    width: 25px;
    margin-left: auto;
    display: block;
    background: transparent;
}

.comic-info {
    gap: 2rem;
    padding: 2rem;
    align-items: flex-start;
}

.comic-info .comic-title {
    font-size: 2rem;
    margin-bottom: 1rem;
    display: block;
}

.comic-info .comic-description {
    line-height: 1.7rem;
}

.comic-detail .options {
    width: 100%;
    display: flex;
    height: 80px;
}

.comic-detail .options button {
    width: 50%;
    font-weight: 700;
    text-transform: uppercase;
    gap: .6rem;
    font-size: 1rem;
}

.comic-detail .options button:hover {
    filter: brightness(103%);
    transition: .1s;
}

.comic-detail .options button img {
    width: auto;
}

.comic-detail .options button.remove-favorite {
    background: var(--secondary);
    color: var(--primary);
}

.comic-detail .options button.buy {
    background: var(--surface-darker);
    color: var(--primary);
}

.comic-detail .options button span {
    color: inherit;
}
