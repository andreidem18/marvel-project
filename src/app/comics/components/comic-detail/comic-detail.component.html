
<shared-modal-wrapper [isOpen]="isOpen" (isOpenChange)="close()" style="width: 80%;"  >

    <div class="comic-detail" *ngIf="comic">

            
        <div class="flex comic-info">
            <img 
                [src]="comicImage" 
                [alt]="comic.title" 
                class="col-4" 
            >

            <div class="details" class="col-8">

                <strong class="comic-title">
                    {{comic.title}}
                </strong>
                <div class="comic-description" [innerHtml]="comic.description">
                </div>
            </div>

            <button class="btn-close" (click)="close()">
                <img 
                    src="../../../../assets/icons/btn-close.png" 
                    alt="close"
                >
            </button>
        </div>

        <div class="options">
            <ng-container *ngIf="!isFavorite; else removeFavorite">
                <button class="add-favorite center" (click)="addToFavorites()">
                    <img src="../../../../assets/icons/btn-favourites-default.png" alt="favorites-icon">
                    <span>Add to favorites</span>
                </button>
            </ng-container>
            <ng-template #removeFavorite>
                <button class="remove-favorite center" (click)="onRemoveFavorite()">
                    <img src="../../../../assets/icons/btn-favourites-primary.png" alt="favorites-icon">
                    <span>Remove from favorites</span>
                </button>
            </ng-template>
            
            <button class="buy center" (click)="buyComic()">
                <img src="../../../../assets//icons/shopping-cart-primary.png" alt="shopping cart icon">
                <span>Buy for ${{comic.prices[0].price}}</span>
            </button>
        </div>

    </div>
</shared-modal-wrapper>
