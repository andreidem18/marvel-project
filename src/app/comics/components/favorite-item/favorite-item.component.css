.favorite-item {
    width: 100%;
    max-width: 240px;
    margin: 0 auto 2rem;
    cursor: pointer;
}

.favorite-item .trash-btn {
    width: 33px;
    background: none;
    margin-bottom: -20px;
    margin-right: -15px;
    margin-left: auto;
    display: block;
    position: relative;
}

.favorite-item .comic-image {
    margin-bottom: .6rem;
}

.favorite-item .comic-name {
    display: block;
    text-align: center;
}


