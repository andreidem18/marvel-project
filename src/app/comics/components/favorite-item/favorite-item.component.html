<div class="favorite-item" (click)="openComicDetail()">
    <button class="trash-btn" (click)="onRemoveFavorite($event)">
        <img src="../../../../assets/icons/btn-delete.png" alt="">
    </button>
    <img 
        [src]="favorite.image" 
        [alt]="favorite.name"
        class="comic-image"
    >
    <strong class="comic-name">
        {{favorite.name}}
    </strong>
</div>

<comic-detail 
    [(isOpen)]="isComicDetailOpen" 
    [comicData]="favorite"
/>
