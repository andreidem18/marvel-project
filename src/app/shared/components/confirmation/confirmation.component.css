.confirmation {
    min-width: 300px;
    max-width: 500px;
}

.message-container {
    padding: 30px;
    text-align: center;
}

.buttons-container {
    display: flex;
    height: 60px;
}

.buttons-container button {
    width: 50%;
    font-size: 1.1rem;
    font-weight: 700;
    /* color: var(--primary); */
    background: var(--surface-dark);
}

.buttons-container button:hover {
    filter: brightness(103%);
    transition: .1s;
}

.buttons-container button.no {
    color: var(--font-color-contrast);
    background: var(--secondary);
}
