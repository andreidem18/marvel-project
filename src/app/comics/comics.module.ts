import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

import { SharedModule } from '../shared/shared.module';

import { ComicDetailComponent } from './components/comic-detail/comic-detail.component';
import { FavoritesListComponent } from './components/favorites-list/favorites-list.component';
import { FavoriteItemComponent } from './components/favorite-item/favorite-item.component';



@NgModule({
  declarations: [
    ComicDetailComponent,
    FavoritesListComponent,
    FavoriteItemComponent,
  ],
  imports: [
    CommonModule,

    SharedModule,
  ],
  exports: [
    ComicDetailComponent,
    FavoritesListComponent,
  ]
})
export class ComicsModule { }
