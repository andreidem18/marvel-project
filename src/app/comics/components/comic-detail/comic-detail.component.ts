import { Component, Input, OnChanges, SimpleChanges, inject, OnInit, Output, EventEmitter } from '@angular/core';
import { ComicsItem } from 'src/app/heroes/interfaces/Hero';
import { Comic } from '../../interfaces/Comic';
import { ComicsService } from '../../services/comics.service';
import { ToastService } from '../../../shared/services/toast/toast.service';
import { ConfirmationService } from '../../../shared/services/confirmation/confirmation.service';

@Component({
  selector: 'comic-detail',
  templateUrl: './comic-detail.component.html',
  styleUrls: ['./comic-detail.component.css']
})
export class ComicDetailComponent implements OnChanges {

  @Input()
  public comicData: ComicsItem | null = null;

  @Input()
  public isOpen: boolean = true;

  @Output()
  public isOpenChange = new EventEmitter<boolean>();

  public comic?: Comic;

  get comicImage(): string {
    const { path, extension } = this.comic?.images[0] || {};
    return `${path}.${extension}`
  }

  get isFavorite(): boolean {
    return this.comicsService.favoritesList
      .some(f => f.name === this.comic?.title)
  }

  private comicsService = inject(ComicsService);
  private toastService = inject(ToastService);
  private confirmationService = inject(ConfirmationService);

  ngOnChanges(changes: SimpleChanges): void {
    const comicData = changes['comicData']?.currentValue as ComicsItem;
    if(!comicData) return;
    this.getComic(comicData)
  }

  getComic(comicData: ComicsItem) {
    this.comic = undefined;
    this.comicsService.getComic(comicData.resourceURI)
      .subscribe(res => {
        this.comic = res.data.results[0];
      });
  }

  close() { 
    this.isOpen = false;
    this.isOpenChange.emit(false);
  }

  addToFavorites() {
    if(!this.comic) return;
    this.comicsService.addToFavorite(this.comic);
    this.toastService.addToast({
      severity: 'success',
      message: 'comic added to favorites!',
      duration: 3,
    })
  }

  onRemoveFavorite() {
    if(!this.comic) return;
    this.confirmationService.display({
      message: `<p>Are you sure you want to 
        remove <b>${this.comic.title}</b> from your favorites?<p>`,
      action: () => {
        this.comicsService.removeFavorite(this.comic!.title);
        this.toastService.addToast({
          message: "Favorite removed successfully",
          duration: 2,
          severity: "success",
        })
      }
    })
  }

  buyComic() {
    this.toastService.addToast({
      message: "this feature haven't been added yet",
      severity: "warning",
      duration: 3,
    })
  }
}
