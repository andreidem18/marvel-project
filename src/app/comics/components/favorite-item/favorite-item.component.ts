import { Component, OnInit, Input, inject } from '@angular/core';
import { Favorite } from '../../interfaces/Favorite';
import { ComicsService } from '../../services/comics.service';
import { ConfirmationService } from 'src/app/shared/services/confirmation/confirmation.service';
import { ToastService } from '../../../shared/services/toast/toast.service';

@Component({
  selector: 'comic-favorite-item',
  templateUrl: './favorite-item.component.html',
  styleUrls: ['./favorite-item.component.css']
})
export class FavoriteItemComponent implements OnInit {

  @Input()
  public favorite!: Favorite;

  public isComicDetailOpen: boolean = false;

  private comicsService = inject(ComicsService);
  private toastService = inject(ToastService);
  private confirmationService = inject(ConfirmationService);

  ngOnInit(): void {
    if(!this.favorite) throw "You must send a favorite item"
  }

  onRemoveFavorite(e: Event) {
    e.stopPropagation();
    this.confirmationService.display({
      message: `<p>Are you sure you want to 
        remove <b>${this.favorite.name}</b> from your favorites?<p>`,
      action: () => {
        this.comicsService.removeFavorite(this.favorite.name);
        this.toastService.addToast({
          message: "Favorite removed successfully",
          duration: 2,
          severity: "success",
        })
      }
    })
  }

  openComicDetail() {
    this.isComicDetailOpen = true;
  }
}
