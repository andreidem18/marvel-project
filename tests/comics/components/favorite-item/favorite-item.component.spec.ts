import { ComponentFixture, TestBed } from '@angular/core/testing';

import { FavoriteItemComponent } from '../../../../src/app/comics/components/favorite-item/favorite-item.component';
import { HttpClientTestingModule } from '@angular/common/http/testing';
import { favoritesMock } from '../../mocks-data/favorites.mock';
import { ComicDetailComponent } from 'src/app/comics/components/comic-detail/comic-detail.component';
import { SharedModule } from 'src/app/shared/shared.module';
import { ComicsService } from 'src/app/comics/services/comics.service';
import { ToastService } from '../../../../src/app/shared/services/toast/toast.service';
import { ConfirmationService } from 'src/app/shared/services/confirmation/confirmation.service';

describe('FavoriteItemComponent', () => {
  const favorite = favoritesMock[0];

  let component: FavoriteItemComponent;
  let fixture: ComponentFixture<FavoriteItemComponent>;
  let compiled: HTMLElement;
  let service: ComicsService;
  let confirmationService: ConfirmationService;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [ HttpClientTestingModule, SharedModule ],
      declarations: [FavoriteItemComponent, ComicDetailComponent],
    });
    fixture = TestBed.createComponent(FavoriteItemComponent);
    service = TestBed.inject(ComicsService);
    confirmationService = TestBed.inject(ConfirmationService);
    component = fixture.componentInstance;
    component.favorite = favorite;
    compiled = fixture.nativeElement;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match with the snapshot', () => {
    expect(compiled.innerHTML).toMatchSnapshot();
  });

  it('should delete favorite when clicking trash button', () => {
    jest.spyOn(confirmationService, 'display');
    const buttonTag = compiled.querySelector('button');
    buttonTag?.dispatchEvent(new Event('click'));
    expect(confirmationService.display).toHaveBeenCalled();
  });

  it('should open detail when clicking favorite', () => {
    const favoriteTag = compiled.querySelector('.favorite-item');
    favoriteTag?.dispatchEvent(new Event('click'));
    expect(component.isComicDetailOpen).toBeTruthy();
  })

});
