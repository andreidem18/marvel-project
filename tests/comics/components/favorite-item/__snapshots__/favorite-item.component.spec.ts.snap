// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteItemComponent should match with the snapshot 1`] = `"<div class="favorite-item"><button class="trash-btn"><img src="../../../../assets/icons/btn-delete.png" alt=""></button><img class="comic-image" src="http://i.annihil.us/u/prod/marvel/i/mg/9/10/591deb1bd1554.jpg" alt="Uncanny X-Men (1963) #405"><strong class="comic-name"> Uncanny X-Men (1963) #405 </strong></div><comic-detail ng-reflect-is-open="false" ng-reflect-comic-data="[object Object]"><shared-modal-wrapper style="width: 80%;" ng-reflect-is-open="false"><div class="modal-container wrap center" ng-reflect-ng-class="[object Object]"><div class="modal-body"><!--bindings={}--></div><div class="overlay wrap"></div></div></shared-modal-wrapper></comic-detail>"`;
